net: tcp: Set correct source IP address when connecting

Frauschi opened the pull request:
In case a TCP connection should be established with a specific source IP address (e.g., by calling bind() prior to connect() on the socket), we have to actually place the address in the outgoing packet.

Fixes #61293

nashif commented:
Please edit  the title and provide something a bit more meaningful.

Frauschi commented:
Hi @nashif
Sorry for the meaningless title, this is my first PR :) I hope the updated one is more suitable.

kartben commented:
> Hi @nashif Sorry for the meaningless title, this is my first PR :) I hope the updated one is more suitable.

Thanks for it! Please also double check the contributing guidelines, especially regarding commit message and sign-off. https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines

I have just triggered a CI build but I can already tell you our compliance check script will complain :)

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -3065,10 +3065,19 @@ int net_tcp_connect(struct net_context *context,
 			net_context_get_iface(context),
 			&net_sin(remote_addr)->sin_addr);
-		conn->src.sin.sin_addr = *ip4;
+			conn->src.sin.sin_addr = *ip4;
```
Please use `net_ipaddr_copy()` here.

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -3088,10 +3097,19 @@ int net_tcp_connect(struct net_context *context,
 					net_context_get_iface(context),
 					&net_sin6(remote_addr)->sin6_addr);
-		conn->src.sin6.sin6_addr = *ip6;
+			conn->src.sin6.sin6_addr = *ip6;
```
Same here.

jukkar reviewed (commented):
Looks good, just make sure the {} are in correct places so that checkpatch is happy.

jukkar commented:
Please change the commit subject to `net: tcp: Set correct source IP address when connecting` and add signed-off-by

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -3065,10 +3065,19 @@ int net_tcp_connect(struct net_context *context,
+			&net_sin(local_addr)->sin_addr))
+		{
+			ip4 = net_if_ipv4_select_src_addr(
 			net_context_get_iface(context),
 			&net_sin(remote_addr)->sin_addr);
```
Those lines should get one extra line of indentation since you've indented `net_if_ipv4_select_src_addr()`, currently the function arguments blend with the code.

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -3065,10 +3065,19 @@ int net_tcp_connect(struct net_context *context,
+		}
+		else
+		{
+			net_ipaddr_copy(&conn->src.sin.sin_addr,
+				&net_sin(local_addr)->sin_addr);
```
nit, but it's good to align arguments  to the opening brackets (like in the `net_ipaddr_copy()` just below).

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -3088,10 +3097,19 @@ int net_tcp_connect(struct net_context *context,
+		}
+		else
+		{
+			net_ipaddr_copy(&conn->src.sin6.sin6_addr,
+				&net_sin6(local_addr)->sin6_addr);
```
nit, but it's good to align arguments to the opening brackets (like in the `net_ipaddr_copy()` just below).

rlubos reviewed (commented):
Thanks @Frauschi for the fix. Please follow Compliance and review suggestions, and we should be good to go :+1:

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -3065,12 +3065,18 @@ int net_tcp_connect(struct net_context *context,
+		}
 		net_ipaddr_copy(&conn->dst.sin.sin_addr,
-				&net_sin(remote_addr)->sin_addr);
+						&net_sin(remote_addr)->sin_addr);
```
Please don't introduce unnecessary changes.

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -3065,12 +3065,18 @@ int net_tcp_connect(struct net_context *context,
+			net_ipaddr_copy(&conn->src.sin.sin_addr, ip4);
+		} else {
+			net_ipaddr_copy(&conn->src.sin.sin_addr,
+							&net_sin(local_addr)->sin_addr);
```
This looks like editor miscofiguration on your side (note that Zephyr uses 8spaces/tab)

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -3088,12 +3094,18 @@ int net_tcp_connect(struct net_context *context,
+		}
 		net_ipaddr_copy(&conn->dst.sin6.sin6_addr,
-				&net_sin6(remote_addr)->sin6_addr);
+						&net_sin6(remote_addr)->sin6_addr);
```
Please don't introduce unnecessary changes.

rlubos reviewed (commented):
Please squash the two commits.

Frauschi commented:
Hi @all. Thanks for the thorough feedback, really appreciate it. I squashed my commits into one and corrected the commit message. I hope we are good to go now.

Frauschi committed:
net: tcp: Set correct source IP address when connecting

In case a TCP connection should be established with a specific source
IP address (e.g., by calling bind() prior to connect() on the socket),
we have to actually place the address in the outgoing packet.

Signed-off-by: Tobias Frauenschläger <t.frauenschlaeger@me.com>

rlubos reviewed (approved):
LGTM (CI issues does not seem to be related to this PR, will need a rerun when fixed on main).
